datadog_checks_base/datadog_checks/base/checks/win/wmi/sampler.py

```diff
@@ -157,15 +158,60 @@ def _query_sample_loop(self):
             self.logger.info("exception in CoInitialize: %s", e)
             raise
 
+        additional_args = []
+        if self.provider != ProviderArchitecture.DEFAULT:
+            context = Dispatch("WbemScripting.SWbemNamedValueSet")
+            context.Add("__ProviderArchitecture", self.provider)
+            additional_args = [None, "", 128, context]
+
+        locator = Dispatch("WbemScripting.SWbemLocator")
+
         while True:
             self._runSampleEvent.wait()
+            if self._stopping:
+                self.logger.debug("_query_sample_loop stopping")
+                return
+
             self._runSampleEvent.clear()
-            if self.is_raw_perf_class and not self._previous_sample:
-                self._current_sample = self._query()
 
-            self._previous_sample = self._current_sample
-            self._current_sample = self._query()
-            self._sampleCompleteEvent.set()
+            try:
+                # Create a new WMI connection.
+                self.logger.debug(
+                    u"Connecting to WMI server (host=%s, namespace=%s, provider=%s, username=%s).",
+                    self.host,
+                    self.namespace,
+                    self.provider,
+                    self.username,
+                )
+                connection = locator.ConnectServer(
+                    self.host, self.namespace, self.username, self.password, *additional_args
+                )
+
+                if self.is_raw_perf_class and not self._previous_sample:
+                    self._current_sample = self._query(connection)
+
+                self._previous_sample = self._current_sample
+                self._current_sample = self._query(connection)
+
+            except Exception as e:
+                self.logger.error("exception in _query_sample_loop: %s", e)
+            finally:
+                connection = None
+                self._sampleCompleteEvent.set()
+
+    def __enter__(self):
+        """
+        I am the context manager
+        """
+        return self
```

Comment:
Should we actually start the thread *here*, instead of in `__init__()`?

```suggestion
        # Moved from `__init__()`:
        self.thread = Thread(target=self._query_sample_loop, name=class_name)
        self.thread.daemon = True
        self.thread.start()
        return self
```

As we switch to a context-managed interface we probably want to make sure nothing starts until we actually enter the block (i.e. instantiating a `WMISampler` shouldn't start anything *yet*).

Reply:
Done, although it means the other usage of WMISampler, which caches an instance, has to call `__enter__` explicitly I couldn't find a "proper" way of doing this (and calling `__exit__`. .Net's `IDisposable` pattern would work better here.

Reply:
Yup, I refactored this a bit to use a cleaner `.start()`/`.stop()` pair of methods that we call into, instead of `.__enter__()`/`.__exit__()`, in 8fbf14b.